crates/agglayer-storage/src/storage/migration/mod.rs

```diff
@@ -0,0 +1,250 @@
+        options.create_missing_column_families(true);
+
+        Ok(DB {
+            rocksdb: rocksdb::DB::open_cf(&options, path, cfs.iter().map(AsRef::as_ref))?,
+            default_write_options: Some(Self::writeopts()),
+        })
+    }
+
+    pub fn create_cfs<S: AsRef<str>>(
+        self,
+        cfs: impl IntoIterator<Item = S>,
+    ) -> Result<Self, DBOpenError> {
+        Ok(self.perform_step(move |db| {
+            for cf in cfs {
+                let cf = cf.as_ref();
+                debug!("Creating column family {cf:?}");
+                db.db
+                    .rocksdb
+                    .create_cf(cf, &rocksdb::Options::default())
+                    .map_err(DBError::from)?;
+            }
+            Ok(())
+        })?)
+    }
+
+    pub fn drop_cfs<S: AsRef<str>>(
+        self,
+        cfs: impl IntoIterator<Item = S>,
+    ) -> Result<Self, DBOpenError> {
+        Ok(self.perform_step(move |db| {
+            for cf in cfs {
+                let cf = cf.as_ref();
+                debug!("Dropping column family {cf:?}");
+                db.db.rocksdb.drop_cf(cf).map_err(DBError::from)?;
+            }
+            Ok(())
+        })?)
+    }
+
+    pub fn migrate(
+        self,
+        migrate_fn: impl FnOnce(&mut DB) -> Result<(), DBMigrationErrorDetails>,
+    ) -> Result<Self, DBOpenError> {
+        Ok(self.perform_step(move |db| {
+            debug!("Running a custom migration step");
+            migrate_fn(&mut db.db)
+        })?)
+    }
+
+    pub fn finalize<'a>(
+        self,
+        expected_schema: impl IntoIterator<Item = &'a str>,
+    ) -> Result<DB, DBOpenError> {
+        if self.step < self.start_step {
+            return Err(DBOpenError::FewerStepsDeclared {
+                declared: self.step,
+                recorded: self.start_step,
+            });
+        }
+
+        let expected_schema: BTreeSet<_> = expected_schema.into_iter().collect();
+        warn!(
+            ?expected_schema,
+            "Expected database schema checking not yet implemented"
+        );
+
+        Ok(self.db)
+    }
+
+    fn perform_step(
+        self,
+        step_fn: impl FnOnce(&mut Self) -> Result<(), DBMigrationErrorDetails>,
+    ) -> Result<Self, DBMigrationError> {
+        let step = self.step;
+        self.perform_step_impl(step_fn)
+            .map_err(|details| DBMigrationError { step, details })
+    }
+
+    #[instrument(skip(self, step_fn), fields(step = self.step))]
+    fn perform_step_impl(
+        mut self,
+        step_fn: impl FnOnce(&mut Self) -> Result<(), DBMigrationErrorDetails>,
+    ) -> Result<Self, DBMigrationErrorDetails> {
+        let step = self.step;
+
+        if step >= self.start_step {
+            info!("Running migration step {step}");
+            step_fn(&mut self)?;
+            self.db.put::<MigrationRecordColumn>(&self.step, &())?;
+        } else {
+            debug!("Step already recorded, skipping");
+        }
+
+        self.step += 1;
```

Comment:
Maybe you could move `self.step` management to `perform_step` function. Or move whole `perform_step_impl` implementation to `perform_step`?

Reply:
> Maybe you could move `self.step` management to `perform_step` function. Or move whole `perform_step_impl` implementation to `perform_step`?

The idea here is for `_impl` to do the work and return a `ErrorDetails` on failure. `perform_step` just adds the step number to the error. This allows to use `?` in the `_impl` method. Otherwise the error handling is a bit too bureaucratic.